lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,444 @@
+    UNEXPECTED_USER_LINKS = "unexpected_user_links"
+    WRONG_USER_LINKS = "wrong_user_links"
+    INCONCEIVABLE = "inconceivable"
+    ERROR = "error"
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+        self.controllers = list()
+        self.bad_controllers = list()
+        self.validation_list = None
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.bad_controllers.append(controller)
+
+    def add_error_or_inconceivable_entries(self, val, controller, dirs):
+        self.validation_list[val][controller] = dirs
+
+    def add_unexpected_entries(self, val, controller, dirs):
+        """Gather values for each validation_list dict key"""
+        self.validation_list[val][controller] = dirs
+
+    def header(self, fp, status):
+        """Adds main starting and ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller):
+        """Checks whether there are controllers in dictionary or not"""
+        return any(
+            [
+                controller in self.validation_list[key]._dict
+                for key in self.validation_list
+            ]
+        )
+
+    """
+    ##FIXME: The mention of hierarchies(archive, controller) from here on, in
+            Hierarchy class is mainly to keep the format algned with the shell script
+    """
+
+    def dump_check(self, fp, controller, key, cnt, hierarchy=False):
+        """Output Messages for each list of controllers, files,
+        directories, etc depicting their cause of being added to
+        that list or dictionary
+        """
+        lead_asterisk = "\t* " if hierarchy == "archive" else "\t"
+        if controller in self.validation_list[key]._dict:
+            if key not in ["subdir_status_indicators", "prefix_status_indicators"]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                if hierarchy == "archive":
+                    cnt = self.output_format(
+                        fp, self.validation_list[key]._dict[controller], cnt, "archive"
+                    )
+                else:
+                    cnt = self.output_format(
+                        fp, self.validation_list[key]._dict[controller], cnt
+                    )
+            elif "subdirs" in self.validation_list[key]._dict[controller]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                cnt += 1
+            elif "prefix_dir" in self.validation_list[key]._dict[controller]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                cnt += 1
```

Comment:
would we really want our return value to be true/False instead of numeric?

Reply:
A function should return whatever type and value is appropriate.  And, it should be named in a way which reflects its purpose and, best case, makes the source code read something like English.

For this case (which I _think_ is `dump_check()`), the function looks through the hierarchy for problems which have been recorded, prints out what it finds, and then returns a value indicating what it did.  I think all we care about in terms of results is whether it found any issues or not -- I don't think we care how many issues it found -- and so returning a boolean value (either `True` or `False`) would be sufficient:
```
    found_issues = False
[...]
                for key in self.validation_list:
                    if self.dump_check([...])
                        found_issues = True
[...]
    return found_issues
```
There is nothing for each layer to count -- it's just a simple question of whether there were any issues encountered.

If we change the function to return a boolean, we should consider renaming it, as well.  This case is a little challenging, because the function does double duty -- it detects the issues _and_ it prints them.  If it only detected them, then a name like 'is_audit_failure()` would be good, such that the code
```
                    if self.is_audit_failure([...])
                        found_issues = True
```
 reads like "if [there] is [an] audit failure [then set] `found_issues` [to] `True`".

We could name it `is_audit_failure_and_print_it()`, but that's unpleasantly long and doesn't actually read well.  So, we can go with something like `print_audit_failures()`, which is nicely expressive, even if it doesn't read as well, and it will (hopefully) be obvious from the code what the return value means, in no small part because we named the return value variable well.

And, note that this idiom doesn't require us to pass in the initial return value -- it doesn't require us to take into account previous values of `found_issues`:  all we care about is whether we found _one_, so if a lower layer returns `True` then we set `found_issues` to `True` (whether it was already `True` or not), and then we return the resulting value to the next layer up when we're done.  So, it's nice, simple, and clean.